### PR TITLE
Added timeout variants for blocking promise ops.

### PR DESCRIPTION
**Pull request by ConnorDoyle**
Supplying a timeout when awaiting completion of an asynchronous event can help avoid "leaking" goroutines that can never make progress.

### CODE REVIEW

**Comment by ConnorDoyle** [comment]
Rebased.

File: pkg/promise/promise.go [hunk 1]
```diff
@@ -49,16 +65,47 @@ func (p *promise) Complete(errors []error) {
...
+	if duration.Nanoseconds() > 0 {
+		timeoutChan = time.After(duration)
+	}
+
+	select {
+	case <-p.completeChan:
+		return p.errors
+	case <-timeoutChan:
+		msg := fmt.Sprintf("Await timed out for promise after [%s]", duration)
+		return []error{NewTimeoutError(msg)}
```
Comment (danielscottt) [inline pkg/promise/promise.go]: I think you may want to do:
```go
return []error{fmt.Errorf("Await timed out for promise after [%s]", duration)}
```
here.

**Comment by danielscottt** [comment]
LGTM mod commend above.

**Comment by ConnorDoyle** [comment]
Restarted build and filed #648 to track that spurious test failure.

File: pkg/promise/promise.go (hunk 1 above)
Comment (ConnorDoyle) [inline pkg/promise/promise.go]: I can make that change.  Admittedly, the custom error type doesn't add much in this case.

Comment (ConnorDoyle) [inline pkg/promise/promise.go]: I will also drop the `TimeoutError` type, since this is the only use site.

**Comment by ConnorDoyle** [comment]
Addressed review comments.

**Pull request merged**